units/host_monitoring_idle_termination.go

```diff
@@ -94,68 +87,147 @@ func (j *idleHostJob) Run(ctx context.Context) {
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "database error for find() by distro ids in [%s]", strings.Join(distroIDsToFind, ",")))
+		return
+	}
+
+	if len(distroIDsToFind) > len(distrosFound) {
+		distroIDsFound := make([]string, 0, len(distrosFound))
+		for _, d := range distrosFound {
+			distroIDsFound = append(distroIDsFound, d.Id)
+		}
+		missingDistroIDs := util.GetSetDifference(distroIDsToFind, distroIDsFound)
+		hosts, err := host.Find(db.Query(host.ByDistroIDs(missingDistroIDs...)))
+		if err != nil {
+			j.AddError(errors.Wrapf(err, "could not find hosts in missing distros: %s", strings.Join(missingDistroIDs, ", ")))
+			return
+		}
+		for _, h := range hosts {
+			j.AddError(errors.Wrapf(h.SetDecommissioned(evergreen.User, "distro is missing"), "could not set host '%s' as decommissioned", h.Id))
+		}
+
+		if j.HasErrors() {
+			return
+		}
+	}
+
+	distrosMap := make(map[string]distro.Distro, len(distrosFound))
+	for i := range distrosFound {
+		d := distrosFound[i]
+		distrosMap[d.Id] = d
+	}
+
+	for _, info := range distroHosts {
+		totalRunningHosts := info.RunningHostsCount
+		minimumHosts := distrosMap[info.DistroID].HostAllocatorSettings.MinimumHosts
+		nIdleHosts := len(info.IdleHosts)
+
+		maxHostsToTerminate := totalRunningHosts - minimumHosts
+		if maxHostsToTerminate <= 0 {
+			continue
+		}
+		nHostsToEvaluateForTermination := nIdleHosts
+		if nIdleHosts > maxHostsToTerminate {
+			nHostsToEvaluateForTermination = maxHostsToTerminate
+		}
+
+		hostsToEvaluateForTermination := make([]host.Host, 0, nHostsToEvaluateForTermination)
+		for i := 0; i < nHostsToEvaluateForTermination; i++ {
+			hostsToEvaluateForTermination = append(hostsToEvaluateForTermination, info.IdleHosts[i])
+			j.AddError(j.checkAndTerminateHost(ctx, &info.IdleHosts[i]))
+		}
+
+		grip.InfoWhen(sometimes.Percent(10), message.Fields{
+			"id":                         j.ID(),
+			"job_type":                   idleHostJobName,
+			"runner":                     "monitor",
+			"op":                         "dispatcher",
+			"distro_id":                  info.DistroID,
+			"minimum_hosts":              minimumHosts,
+			"num_running_hosts":          totalRunningHosts,
+			"num_idle_hosts":             nIdleHosts,
+			"num_idle_hosts_to_evaluate": nHostsToEvaluateForTermination,
+			"idle_hosts_to_evaluate":     hostsToEvaluateForTermination,
+		})
+	}
+}
+
+func (j *idleHostJob) checkAndTerminateHost(ctx context.Context, h *host.Host) error {
+	if !h.IsEphemeral() {
 		grip.Notice(message.Fields{
 			"job":      j.ID(),
-			"host_id":  j.HostID,
+			"host_id":  h.Id,
 			"job_type": j.Type().Name,
-			"status":   j.host.Status,
-			"provider": j.host.Distro.Provider,
+			"status":   h.Status,
+			"provider": h.Distro.Provider,
 			"message":  "host termination for a non-spawnable distro",
 			"cause":    "programmer error",
 		})
-		j.AddError(errors.New("non-spawnable host"))
-		return
+		return errors.New("non-spawnable host")
```

Comment:
I know you didn't write it, but this is a weird error message. Can this be "attempted to idle terminate non-ephemeral host"?